airbyte-integrations/connectors/destination-clickhouse/src/main/java/io/airbyte/integrations/destination/clickhouse/ClickhouseSqlOperations.java

```diff
@@ -0,0 +1,95 @@
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.sql.SQLException;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ru.yandex.clickhouse.ClickHouseConnection;
+import ru.yandex.clickhouse.ClickHouseStatement;
+import ru.yandex.clickhouse.domain.ClickHouseFormat;
+
+public class ClickhouseSqlOperations extends JdbcSqlOperations {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ClickhouseSqlOperations.class);
+
+  @Override
+  public void createSchemaIfNotExists(final JdbcDatabase database, final String schemaName) throws Exception {
+    database.execute(String.format("CREATE DATABASE IF NOT EXISTS %s;\n", schemaName));
+  }
+
+  @Override
+  public String createTableQuery(final JdbcDatabase database, final String schemaName, final String tableName) {
+    return String.format(
+        "CREATE TABLE IF NOT EXISTS %s.%s ( \n"
+            + "%s String,\n"
+            + "%s String,\n"
+            + "%s DateTime64(3, 'GMT') DEFAULT now(),\n"
+            + "PRIMARY KEY(%s)\n"
+            + ")\n"
+            + "ENGINE = MergeTree;\n",
+        schemaName, tableName,
+        JavaBaseConstants.COLUMN_NAME_AB_ID,
+        JavaBaseConstants.COLUMN_NAME_DATA,
+        JavaBaseConstants.COLUMN_NAME_EMITTED_AT,
+        JavaBaseConstants.COLUMN_NAME_AB_ID,
+        JavaBaseConstants.COLUMN_NAME_AB_ID);
+  }
+
+  @Override
+  public void executeTransaction(final JdbcDatabase database, final List<String> queries) throws Exception {
+    final StringBuilder appendedQueries = new StringBuilder();
+    for (final String query : queries) {
+      appendedQueries.append(query);
+    }
+    database.execute(appendedQueries.toString());
+  }
+
+  @Override
+  public void insertRecordsInternal(final JdbcDatabase database,
+                                    final List<AirbyteRecordMessage> records,
+                                    final String schemaName,
+                                    final String tmpTableName)
+      throws SQLException {
+    if (records.isEmpty()) {
+      return;
+    }
+
+    database.execute(connection -> {
+      File tmpFile = null;
+      try {
+        tmpFile = Files.createTempFile(tmpTableName + "-", ".tmp").toFile();
+        writeBatchToFile(tmpFile, records);
+
+        ClickHouseConnection conn = connection.unwrap(ClickHouseConnection.class);
+        ClickHouseStatement sth = conn.createStatement();
+        sth.write() // Write API entrypoint
+            .table(String.format("%s.%s", schemaName, tmpTableName)) // where to write data
+            .data(tmpFile, ClickHouseFormat.CSV) // specify input
+            .send();
+
+      } catch (final Exception e) {
+        throw new RuntimeException(e);
+      } finally {
+        try {
+          if (tmpFile != null) {
+            Files.delete(tmpFile.toPath());
+          }
+        } catch (final IOException e) {
+          throw new RuntimeException(e);
```

Comment:
[Optional] Throwing an exception from within a finally block will mask any exception which was previously thrown in the try or catch block, and the masked’s exception message and stack trace will be lost.

Reply:
thanks, I copied that code from Postgres connector:

https://github.com/airbytehq/airbyte/blob/b5af3f1d561e486c17083ff207973d67113b7a40/airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/PostgresSqlOperations.java#L54

If you have better resolution, I'd happy to change.

Reply:
How about this:   (I'm implementing similar pattern for Destination MariaDB ColumnStore now)
```java
Exception primaryException = null;
try {
  // your code
} catch (final Exception e) {
  primaryException = e;
  throw new RuntimeException(primaryException);
} finally {
  try {
    // your code
  } catch (final IOException e) {
    if (primaryException != null) e.addSuppressed(primaryException);
    throw new RuntimeException(e);
  }
}
```

Reply:
thanks @koji-m, that's great.